# 0. Presentación

En este cuaderno se realiza la etapa del ***Proceso KDD de Selección de Datos*** en este caso, aplicándola sobre los diferentes conjuntos de datos abiertos de la OMS, para el análisis de datos en su vertiente de estudiar cuales son las causas de la expansión de las enfermedades infecciosas en la sociedad actual.

In [ ]:
# Carga y manipulación de datos
from google.colab import files
import io
 
# Librería para manipular DataFrames
import pandas as pd

# 1. Creación de dataframe vacío

Creamos un dataframe vacío para ir almacenando en él los datos correspondientes.

In [ ]:
df = pd.DataFrame()

Ahora, se va a proceder a trabajar con los diferentes datasets del Data Lake, con el objetivo de encontrar una tarjeta de datos que sirva como entrada para los algoritmos a realizar después. Debido a la gran cantidad de bases de datos, se le asignarán unos identificadores para trabajar de forma organizada.

# 2. Fallecidos a causa de enfermedades infecciosas durante la infancia (enf_inf)





In [ ]:
## Carga de Datos
enf_inf_1 = pd.read_csv('Infecciones_Agudas_Vias_Respiratorias.csv', sep=',') ## Infecciones agudas en las vías respiratorias
enf_inf_2 = pd.read_csv('Sepsis.csv', sep=',') ## Sepsis
enf_inf_3 = pd.read_csv('Meningitis_encefalitis.csv', sep=',') ## Meningitis y encefalitis
enf_inf_4 = pd.read_csv('Malaria.csv', sep=',') ## Malaria
enf_inf_5 = pd.read_csv('Sarampion.csv', sep=',') ## Sarampión
enf_inf_6 = pd.read_csv('VIH.csv', sep=',') ## VIH
enf_inf_7 = pd.read_csv('Diarrea.csv', sep=',') ## Diarrea

In [ ]:
def preproceso_enf(dataset):
  dataset = dataset.drop(['2000.1','2001.1','2002.1','2003.1','2004.1','2005.1','2006.1','2007.1','2008.1',
                      '2009.1','2010.1','2011.1','2012.1','2013.1','2014.1','2015.1','2016.1','2017.1',
                      '2000','2001','2002','2003','2004','2005','2006','2007','2008',
                      '2009','2010','2011','2012','2013','2014','2015','2016','2017'], axis=1)
  dataset = dataset.drop([0,1,2], axis=0)

  for column in dataset:
    if column != "Unnamed: 0":
      dataset[column] = dataset[column].astype(str).astype(int)

  num_years = len(dataset.columns) - 1 # Columna con nombre de países.
  dataset['Total'] = dataset.sum(axis=1)
  dataset['Media'] = dataset['Total'] / num_years
  
  return dataset

enf_inf_1 = preproceso_enf(enf_inf_1)
enf_inf_2 = preproceso_enf(enf_inf_2)
enf_inf_3 = preproceso_enf(enf_inf_3)
enf_inf_4 = preproceso_enf(enf_inf_4)
enf_inf_5 = preproceso_enf(enf_inf_5)
enf_inf_6 = preproceso_enf(enf_inf_6)
enf_inf_7 = preproceso_enf(enf_inf_7)

In [ ]:
### Al tener el mismo número de países y ordenados igual, para rellenar los datos, basta con hacer lo siguiente:
df['Pais'] = enf_inf_1['Unnamed: 0']
df['Total Fallecidos Vias Respiratorias'] = enf_inf_1['Total']
df['Media Fallecidos Vias Respiratorias'] = enf_inf_1['Media']
df['Total Fallecidos Sepsis'] = enf_inf_2['Total']
df['Media Fallecidos Sepsis'] = enf_inf_2['Media']
df['Total Fallecidos Meningitis Encefalitis'] = enf_inf_3['Total']
df['Media Fallecidos Meningitis Encefalitis'] = enf_inf_3['Media']
df['Total Fallecidos Malaria'] = enf_inf_4['Total']
df['Media Fallecidos Malaria'] = enf_inf_4['Media']
df['Total Fallecidos Sarampion'] = enf_inf_5['Total']
df['Media Fallecidos Sarampion'] = enf_inf_5['Media']
df['Total Fallecidos VIH'] = enf_inf_6['Total']
df['Media Fallecidos VIH'] = enf_inf_6['Media']
df['Total Fallecidos Diarrea'] = enf_inf_7['Total']
df['Media Fallecidos Diarrea'] = enf_inf_7['Media']

# 3. Índices de vacunación (vac)

In [ ]:
## Carga de Datos
vac_1 = pd.read_csv('PolioInmunizationCoveragePerCountry.csv', sep=',') ## Polio
vac_2 = pd.read_csv('PneumococcalConjugateInmunizationPerCountry.csv', sep=',') ## Conjugado neumocócico
vac_3 = pd.read_csv('MeaslesFirstDosePerCountry.csv', sep=',') ## Sarampión 1º Dosis
vac_4 = pd.read_csv('MeaslesSecondDosePerCountry.csv', sep=',') ## Sarampión 2º Dosis
vac_5 = pd.read_csv('HepatitisBInmunizationCoveragePerCountry.csv', sep=',') ## Hepatitis B
vac_6 = pd.read_csv('HaemophilusInfluenzaeBInmunizationPerCountry.csv', sep=',') ## Haemophilus influenzae tipo b
vac_7 = pd.read_csv('DiphtheriaTetanusToxoidAndTertussisInmunizationPerCountry.csv', sep=',') ## Difteria
vac_8 = pd.read_csv('BCGInmunizationPerCountry.csv', sep=',') ## Bacilo Calmette-Guérin

In [ ]:
def preproceso_vac(dataset):
  dataset = dataset.drop([0], axis=0)

  num_years = len(dataset.columns)
  dataset['Total'] = dataset.sum(axis=1)
  dataset['Media'] = dataset['Total'] / num_years
  
  return dataset

vac_1 = preproceso_vac(vac_1)
vac_2 = preproceso_vac(vac_2)
vac_3 = preproceso_vac(vac_3)
vac_4 = preproceso_vac(vac_4)
vac_5 = preproceso_vac(vac_5)
vac_6 = preproceso_vac(vac_6)
vac_7 = preproceso_vac(vac_7)
vac_8 = preproceso_vac(vac_8)

In [ ]:
## Al no contar con el mismo número de países y por tanto, cambiar el orden, para rellenar los datos se requiere lo siguiente:
def rellenar_vac(dataset,nombre_columna):
  df[nombre_columna] = 'NaN' # Crear nueva columna con valor predeterminado NaN.
  for i in dataset.index:
    for y in df.index:
      if dataset['Unnamed: 0'][i] == df['Pais'][y]:
        df[nombre_columna][y] = dataset['Media'][i]

  return df

df = rellenar_vac(vac_1,'Vacunación contra la poliomielitis entre los niños de 1 año (%)')
df = rellenar_vac(vac_2,'Vacunación de antineumocócicas conjugadas en niños de 1 año (%)')
df = rellenar_vac(vac_3,'Vacunación de la primera dosis contra el sarampión (%)')
df = rellenar_vac(vac_4,'Vacunación de la segunda dosis contra el sarampión (%)')
df = rellenar_vac(vac_5,'Vacunación contra la hepatitis B entre los niños de 1 año (%)')
df = rellenar_vac(vac_6,'Vacunación contra Hib entre niños de 1 año (%)')
df = rellenar_vac(vac_7,'Vacunación contra la tos ferina y el toxoide tetánico diftérico en niños de 1 año (%)')
df = rellenar_vac(vac_8,'Vacunación con BCG en niños de 1 año (%)')

# 4. Infraestructuras y personal sanitario (inf_per)

In [ ]:
## Carga de Datos
inf_per_1 = pd.read_csv('HealthInfrastructurePerCountry.csv', sep=',') # Infraestructura sanitarias
inf_per_2 = pd.read_csv('HealthWorkerDensityPerCountry.csv', sep=',') # Personal sanitario
inf_per_3 = pd.read_csv('HospistalBedsPerCountry.csv', sep=',') # Camas de hospital

## Infraestructuras sanitarias

In [ ]:
inf_per_1_groupBy = inf_per_1.groupby('Country').mean()
inf_per_1_groupBy

Year  ...  Total density per 100 000 population: Hospitals
Country                      ...                                                 
Afghanistan          2011.5  ...                                             0.37
Albania              2011.5  ...                                             1.36
Angola               2010.0  ...                                              NaN
Antigua and Barbuda  2011.5  ...                                             1.11
Armenia              2011.5  ...                                             4.03
...                     ...  ...                                              ...
Uruguay              2011.5  ...                                             3.93
Vanuatu              2011.5  ...                                             2.37
Yemen                2011.5  ...                                             2.97
Zambia               2011.5  ...                                             0.45
Zimbabwe             2011.5  ...                                             0.52

[144 rows x 7 columns]

In [ ]:
lista_centrossanitarios = []

for i in df.index:
  encontrado = False
  for y in inf_per_1_groupBy.index.values:
    if y == df["Pais"][i]:
      lista_centrossanitarios.append(inf_per_1_groupBy['Total density per 100 000 population: Health centres'][y])
      encontrado = True

  if encontrado == False:
      lista_centrossanitarios.append('NaN')

df['Centros de salud (Densidad por 100000 hab.)'] = lista_centrossanitarios

In [ ]:
lista_hospitales = []

for i in df.index:
  encontrado = False
  for y in inf_per_1_groupBy.index.values:
    if y == df["Pais"][i]:
      lista_hospitales.append(inf_per_1_groupBy['Total density per 100 000 population: Hospitals'][y])
      encontrado = True

  if encontrado == False:
      lista_hospitales.append('NaN')

df['Hospitales (Densidad por 100000 hab.)'] = lista_hospitales

## Personal Sanitario

In [ ]:
inf_per_2_groupBy = inf_per_2.groupby('Country').mean()
inf_per_2_groupBy

Year  ...  Pharmacists  (per 10 000 population)
Country                                          ...                                      
Afghanistan                         2011.312500  ...                              0.339308
Albania                             2003.296296  ...                              4.477278
Algeria                             2010.833333  ...                              2.541250
Andorra                             2002.928571  ...                              9.236786
Angola                              2009.666667  ...                              0.253000
...                                         ...  ...                                   ...
Venezuela (Bolivarian Republic of)  2008.250000  ...                                   NaN
Viet Nam                            2009.076923  ...                              2.640333
Yemen                               2011.000000  ...                              0.582250
Zambia                              2010.545455  ...                              0.359500
Zimbabwe                            2009.055556  ...                              0.333250

[194 rows x 5 columns]

In [ ]:
lista_doctores = []

for i in df.index:
  encontrado = False
  for y in inf_per_2_groupBy.index.values:
    if y == df["Pais"][i]:
      lista_doctores.append(inf_per_2_groupBy['Medical doctors (per 10 000 population)'][y])
      encontrado = True

  if encontrado == False:
      lista_doctores.append('NaN')

df['Doctores medicos (por 10000 hab.)'] = lista_doctores

In [ ]:
lista_enfermeros = []

for i in df.index:
  encontrado = False
  for y in inf_per_2_groupBy.index.values:
    if y == df["Pais"][i]:
      lista_enfermeros.append(inf_per_2_groupBy['Nursing and midwifery personnel (per 10 000 population)'][y])
      encontrado = True

  if encontrado == False:
      lista_enfermeros.append('NaN')

df['Enfermeros y personal de parto (por 10000 hab.)'] = lista_enfermeros

In [ ]:
lista_farmaceuticos = []

for i in df.index:
  encontrado = False
  for y in inf_per_2_groupBy.index.values:
    if y == df["Pais"][i]:
      lista_farmaceuticos.append(inf_per_2_groupBy['Pharmacists  (per 10 000 population)'][y])
      encontrado = True

  if encontrado == False:
      lista_farmaceuticos.append('NaN')

df['Farmaceuticos (por 10000 hab.)'] = lista_farmaceuticos

## Camas de hospital

In [ ]:
inf_per_3_groupBy = inf_per_3.groupby('Country').mean()
inf_per_3_groupBy

Year  Hospital beds (per 10 000 population)
Country                                                                               
Afghanistan                         2008.500000                               4.322222
Albania                             2006.384615                              30.523077
Algeria                             2009.500000                              18.000000
Angola                              2005.000000                               8.000000
Antigua and Barbuda                 2014.333333                              30.983333
...                                         ...                                    ...
Venezuela (Bolivarian Republic of)  2014.600000                               8.340000
Viet Nam                            2007.000000                              26.925000
Yemen                               2008.500000                               6.672222
Zambia                              2009.000000                              19.500000
Zimbabwe                            2008.500000                              23.500000

[178 rows x 2 columns]

In [ ]:
lista_camas = []

for i in df.index:
  encontrado = False
  for y in inf_per_3_groupBy.index.values:
    if y == df["Pais"][i]:
      lista_camas.append(inf_per_3_groupBy['Hospital beds (per 10 000 population)'][y])
      encontrado = True

  if encontrado == False:
      lista_camas.append('NaN')

df['Camas de hospital (por 10000 hab.)'] = lista_camas

# 5. Bases de datos relacionadas con el peso (peso)


In [ ]:
## Carga de Datos
peso_1 = pd.read_csv('LowBirthWeightByCountry.csv', sep=',') # Bajo peso al nacer.
peso_2 = pd.read_csv('Anemia0to5YearsByCountry.csv', sep=',') # Anemia desde los 0 a 5 años.
peso_3 = pd.read_csv('ExclusiveBreastfeedingPerCountry.csv', sep=',') # Leche materna.
peso_4 = pd.read_csv('Thinness5to9YearsPerCountry.csv', sep=',') # Delgadez extrema de 5 a 9 años.

## Bajo peso al nacer

In [ ]:
peso_1_groupBy = peso_1.groupby('Country').mean()
peso_1_groupBy

Year  ...  Low birth weight prevalence (%)
Country                                     ...                                 
Albania                             2007.5  ...                          4.71250
Algeria                             2007.5  ...                          7.45625
Andorra                             2007.5  ...                          7.45625
Angola                              2007.5  ...                         16.62500
Antigua and Barbuda                 2007.5  ...                          9.21875
...                                    ...  ...                              ...
Vanuatu                             2007.5  ...                         11.00000
Venezuela (Bolivarian Republic of)  2007.5  ...                          8.69375
Viet Nam                            2007.5  ...                          8.59375
Zambia                              2007.5  ...                         12.43125
Zimbabwe                            2007.5  ...                         12.68750

[146 rows x 3 columns]

In [ ]:
lista_bajo_peso = []

for i in df.index:
  encontrado = False
  for y in peso_1_groupBy.index.values:
    if y == df["Pais"][i]:
      lista_bajo_peso.append(peso_1_groupBy['Low birth weight prevalence (%)'][y])
      encontrado = True

  if encontrado == False:
      lista_bajo_peso.append('NaN')

df['Niños con bajo peso al nacer(%)'] = lista_bajo_peso

## Anemia desde los 0 a 5 años

In [ ]:
peso_2 = peso_2.drop([0], axis = 0)
peso_2

num_years = len(peso_2) - 1 # Columna con nombre de países.
peso_2['Total'] = peso_2.sum(axis=1)
peso_2['Media'] = peso_2['Total'] / num_years


df['Prevalencia de anemia en niños menores de 5 años (%)'] = 'NaN' # Crear nueva columna con valor predeterminado NaN.
for i in peso_2.index:
  for y in df.index:
    if peso_2['Unnamed: 0'][i] == df['Pais'][y]:
        df['Prevalencia de anemia en niños menores de 5 años (%)'][y] = peso_2['Media'][i]

## Leche Materna

In [ ]:
peso_3_groupBy = peso_3.groupby('Country').mean()
peso_3_groupBy

Infants exclusively breastfed for the first six months of life (%)
Country                                                                                               
Afghanistan                                                                 43.100000                 
Albania                                                                     15.600000                 
Algeria                                                                     14.966667                 
Angola                                                                      37.400000                 
Argentina                                                                   32.000000                 
...                                                                               ...                 
Venezuela (Bolivarian Republic of)                                           7.100000                 
Viet Nam                                                                    18.383333                 
Yemen                                                                       13.025000                 
Zambia                                                                      36.583333                 
Zimbabwe                                                                    27.300000                 

[154 rows x 1 columns]

In [ ]:
lista_leche_materna = []

for i in df.index:
  encontrado = False
  for y in peso_3_groupBy.index.values:
    if y == df["Pais"][i]:
      lista_leche_materna.append(peso_3_groupBy['Infants exclusively breastfed for the first six months of life (%)'][y])
      encontrado = True

  if encontrado == False:
      lista_leche_materna.append('NaN')

df['Niños amamantados en sus primeros 6 meses de vida (%)'] = lista_leche_materna

# 6. Bases de datos relacionadas con contaminación e higiene (cont)

In [ ]:
## Carga de Datos
cont_1 = pd.read_csv('AirPollutionPerCountry.csv', sep=',') # Polución del aire.
cont_2 = pd.read_csv('WaterAndHygienePerCountry.csv', sep=',') # Agua potable insalubre, saneamiento e higiene de manos. 

## Polución del aire

In [ ]:
lista_polucion = []

for i in df.index:
  encontrado = False
  for y in cont_1.index:
    if cont_1["Unnamed: 0"][y] == df["Pais"][i]:
      lista_polucion.append(cont_1['Concentrations of fine particulate matter (PM2.5)'][y])
      encontrado = True

  if encontrado == False:
      lista_polucion.append('NaN')

df['Concentracion de partículas finas (polucion) en el aire'] = lista_polucion

## Agua potable insalubre, saneamiento e higiene de manos

In [ ]:
lista_dalys_higiene = []

for i in df.index:
  encontrado = False
  for y in cont_2.index:
    if cont_2["Unnamed: 0"][y] == df["Pais"][i]:
      lista_dalys_higiene.append(cont_2["Water, sanitation and hygiene attributable DALYs ('000) in children under 5 years"][y])
      encontrado = True

  if encontrado == False:
      lista_dalys_higiene.append('NaN')

df['DALY debido al agua, saneamiento e higiene en niños menores de 5 años'] = lista_dalys_higiene

In [ ]:
lista_muertes_higiene = []

for i in df.index:
  encontrado = False
  for y in cont_2.index:
    if cont_2["Unnamed: 0"][y] == df["Pais"][i]:
      lista_muertes_higiene.append(cont_2["Water, sanitation and hygiene attributable deaths ('000) in children under 5 years"][y])
      encontrado = True

  if encontrado == False:
      lista_muertes_higiene.append('NaN')

df['Muertes debido al agua, saneamiento e higiene en niños menores de 5 años'] = lista_muertes_higiene

# 7. Base de datos resultante

El resultado tras llevar a cabo este proceso de selección de datos se plasma en el siguiente dataframe, que cuenta con 194 registros, que hacen referencia a diferentes países, y 35 columnas.

In [ ]:
df

Pais  ...  Muertes debido al agua, saneamiento e higiene en niños menores de 5 años
3                           Afghanistan  ...                                              62200                       
4                               Albania  ...                                                116                       
5                               Algeria  ...                                               3139                       
6                               Andorra  ...                                                  0                       
7                                Angola  ...                                              37920                       
..                                  ...  ...                                                ...                       
192  Venezuela (Bolivarian Republic of)  ...                                                857                       
193                            Viet Nam  ...                                               4905                       
194                               Yemen  ...                                              12890                       
195                              Zambia  ...                                              10541                       
196                            Zimbabwe  ...                                               4371                       

[194 rows x 35 columns]

# 8. Exportación de datos

Por último, convertimos el dataframe en un archivo CSV, para poder descargarlo.

In [ ]:
df.to_csv('oms_data.csv')